sys/kern/tmpfs.c

```diff
@@ -48,63 +90,170 @@ typedef struct tmpfs_node {
+
   /* Data that is only applicable to a particular type. */
   union {
     struct {
-      struct tmpfs_node *parent;   /* Parent directory. */
-      tmpfs_dirent_list_t dirents; /* List of directory entries. */
+      struct tmpfs_node *parent;    /* Parent directory. */
+      tmpfs_dirent_list_t dirents;  /* List of directory entries. */
+      tmpfs_dirent_list_t fdirents; /* List of free directory entries. */
     } tfn_dir;
-    struct {
-      blkptr_t direct[DIRECT_BLK_NO];
-      blkptr_t *l1indirect;
-      blkptr_t **l2indirect;
-      size_t nblocks; /* Number of blocks used by this file. */
-    } tfn_reg;
     struct {
       char *link;
     } tfn_lnk;
   };
 } tmpfs_node_t;
 
+typedef STAILQ_HEAD(, mem_arena) mem_arenalist_t;
+
 typedef struct tmpfs_mount {
   tmpfs_node_t *tfm_root;
   mtx_t tfm_lock;
   ino_t tfm_next_ino;
+  mem_arenalist_t tfm_arenas;
 } tmpfs_mount_t;
 
-static int alloc_block(tmpfs_mount_t *tfm, tmpfs_node_t *v, blkptr_t *blkptrp) {
-  assert(blkptrp != NULL);
+typedef struct mem_arena {
+  STAILQ_ENTRY(mem_arena) tma_link; /* link on list of all arenas */
+  size_t tma_ninodes;               /* number of free inodes */
+  size_t tma_ndblocks;              /* number of free data blocks */
 
-  if (!(*blkptrp)) {
-    blkptr_t blkptr = kmem_alloc(PAGESIZE, M_ZERO);
-    if (!blkptr)
-      return ENOMEM;
-    v->tfn_reg.nblocks++;
-    *blkptrp = blkptr;
+  /* bitmaps of free items */
+  bitstr_t tma_inode_bm[bitstr_size(ARENA_INODE_CNT)];
+  bitstr_t tma_dblock_bm[bitstr_size(ARENA_DATA_BLOCKS)];
+
+  /* pointers to first item */
+  tmpfs_node_t *tma_inodes;
+  blkptr_t tma_dblocks;
+} mem_arena_t;
+
+static void try_alloc_frame(vaddr_t va) {
+  paddr_t pap;
+  va &= ~(PAGESIZE - 1); /* align address to the page size */
+  if (!pmap_extract(pmap_kernel(), va, &pap))
+    kva_map(va, BLOCK_SIZE, M_ZERO);
+}
+
+/* tmpfs memory allocation routines */
+static void mem_init_arena(mem_arena_t *arena) {
+  arena->tma_ninodes = ARENA_INODE_CNT;
+  arena->tma_ndblocks = ARENA_DATA_BLOCKS;
+
+  bit_nset(arena->tma_inode_bm, 0, ARENA_INODE_CNT - 1);
+  bit_nset(arena->tma_dblock_bm, 0, ARENA_DATA_BLOCKS - 1);
+
+  arena->tma_inodes = (void *)arena + ARENA_HEADER_SIZE;
+  arena->tma_dblocks = (void *)arena + (1 + ARENA_INODE_BLOCKS) * BLOCK_SIZE;
+}
+
+static mem_arena_t *mem_new_arena(mem_arenalist_t *al) {
+  mem_arena_t *newar = (mem_arena_t *)kva_alloc(ARENA_SIZE);
+  if (newar == NULL)
+    return NULL;
+  kva_map((vaddr_t)newar, BLOCK_SIZE, M_ZERO);
+
+  mem_init_arena(newar);
+
+  STAILQ_INSERT_TAIL(al, newar, tma_link);
+  return newar;
+}
+
+static mem_arena_t *mem_find_ptr_arena(mem_arenalist_t *al, void *ptr) {
+  mem_arena_t *arena;
+  STAILQ_FOREACH(arena, al, tma_link) {
+    if ((uintptr_t)ptr > (uintptr_t)arena &&
+        (uintptr_t)ptr < (uintptr_t)arena + ARENA_SIZE) {
+      return arena;
+    }
   }
+  return NULL;
+}
 
-  return 0;
+static blkptr_t mem_arena_alloc_dblk(mem_arena_t *arena) {
```

Comment:
Functions that read and/or modify state that is shared across multiple vnodes must be protected by a lock. This state includes the arena list and the arenas themselves.

Reply:
What an embarrassing oversight on my part. Fixed.

Reply:
It is now protected by the mount point's lock.